### PR TITLE
Emulator works with upstream hotload

### PR DESCRIPTION
**Pull request by vicancy**
Usage: asrs-emulator [options] [command]

Options:
-?|-h|--help Show help information

Commands:
start To start the emulator.
upstream To init/list/update the upstream options

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Emulator/Program.cs [hunk 1]
```diff
@@ -1,35 +1,133 @@
...
+                command.Description = "To init/list/update the upstream options";
+                command.HelpOption("-h|--help");
+                command.Command("init", c =>
+                {
+                    c.Description = "Init the default upstream options into a settings.json config";
+                    c.OnExecute(() =>
+                    {
+                        if (File.Exists(SettingsFile))
+                        {
+                            Console.WriteLine($"Already contains {SettingsFile}, still want to override it with the default one? (N/y)");
```
Comment (KKhurin) [inline src/Microsoft.Azure.SignalR.Emulator/Program.cs:47]: Add Path.GetFullPath here to avoid confusion?

**Review by zackliu — APPROVED** [review]

File: src/Microsoft.Azure.SignalR.Emulator/Program.cs [hunk 2]
```diff
@@ -1,35 +1,192 @@
...
+                {
+                    c.Description = "Init the default upstream options into a settings.json config. Use -o to specify the folder to export the default settings.";
+                    var configOptions = c.Option("-o|--output", "Specify the folder to init the upstream settings file.", CommandOptionType.SingleValue);
+                    c.HelpOption("-h|--help");
+                    c.OnExecute(() =>
+                    {
+                        string outputFile = configOptions.HasValue() ? Path.GetFullPath(Path.Combine(configOptions.Value(), SettingsFileName)) : SettingsFile;
+                        if (File.Exists(outputFile))
+                        {
+                            Console.WriteLine($"Already contains '{outputFile}', still want to override it with the default one? (N/y)");
```
Comment (zackliu) [inline src/Microsoft.Azure.SignalR.Emulator/Program.cs:49]: Why the `N` is upper case while `y` is lower case

**Review by vwxyzh — APPROVED** [review]

**Review by KKhurin — APPROVED** [review]

File: src/Microsoft.Azure.SignalR.Emulator/Program.cs [hunk 3]
```diff
@@ -1,35 +1,192 @@
...
+        private static readonly string SettingsFile = Path.GetFullPath(SettingsFileName);
+        private static readonly string ProgramDefaultSettingsFile = Path.Combine(AppContext.BaseDirectory, SettingsFileName);
+        private static readonly string AppSettingsFile = Path.Combine(AppContext.BaseDirectory, "appsettings.json");
+
         public static void Main(string[] args)
         {
             // todo: "upstream init" to generate the default setting.json file
             // todo: "upstream list" to list the settings file
             // todo: "start" to run the emulator
             // todo: "help" to explain the upstream and pattern rules
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Emulator/Program.cs:28]: can be removed.

**Review by JialinXin — APPROVED** [review]
:shipit:

File: src/Microsoft.Azure.SignalR.Emulator/Program.cs (hunk 2 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Emulator/Program.cs:44]: upper case is the default one

**Pull request merged**